content/blog/2019-10-23-tail-call-elimination.md

```diff
@@ -0,0 +1,239 @@
+call is a call to a function whose value is immediately returned. For example,
+`return foo()` is a tail call. Whenever we have a tail call in our Bril program,
+we do not need to create a new stack frame for it. We can simply "overwrite"
+the current stack frame since won't need any of the values in the frame anymore.
+This is crucial for programming languages that use *tail recursion* as a
+programming idiom (e.g. OCaml, Haskell, etc.). Consider the following (somewhat
+contrived) TypeScript program:
+```
+function loop(n: number) {
+  if (n == 0) {
+    return 0;
+  } else {
+    return loop(n-1);
+  }
+}
+```
+This function simply loops `n` times, but does so in a functional way. Without
+tail call elimination, programs like this would stack overflow with large values
+of `n`. For languages that depend on this idiom, this is unacceptable.
+
+## Design
+The base Bril language is not rich enough to express tail call elimination, or
+even function calls in general. To enrich the language, we can make Bril more
+closely resemble something like x86 assembly. Intuitively, all Bril variables
+implictly live on the stack. To pass arguments in a function call, we must
+explicitly push them onto the stack. This is akin to pushing values on the stack
+in assembly. When a function returns, those arguments are implictly popped off the stack.
+Additionally, the return value of the function is obtained using a special
+keyword that is akin to getting the return value from `rax` according to the
+[System V Calling Conventions](https://en.wikipedia.org/wiki/X86_calling_conventions#x86-64_calling_conventions).
+
+Importantly, we need the capability to jump program locations other than the
+labels in the program. This way, if we have a tail call, then we can jump to
+the beginning of the callee instead of creating a new stack frame. This is [how
+tail call elimination is implemented](https://www.cs.cornell.edu/courses/cs4120/2019sp/lectures/34functional/lec34-sp19.pdf?1556325712)
+in x86 assembly.
+
+We have to modify the grammar to support features like pushing onto the stack
+and defining functions, then update the interpreter. Thanks to the work done by
+Alexa and Greg for [Project 1](https://github.com/sampsyo/bril/pull/16), these
+changes were much easier for me to make.
+
+## Implementation
+
+### Modifying the Grammar
+The first step is to modify the grammar to support function declarations with
+arguments and return types, as well as support the various new value/effect
+operations. Functions look as follows:
+```
+int foo(x: int, b: bool) {
+  print x;
+  print b;
+  ...
+}
+```
+Functions must specify a return type, a name, and a potentially empty list of
+typed arguments.
+
+Here are the new effect operations and their semantics:
+- `push arg1 ... argn`: push arguments to the stack, which can be used by the
+first function to be `call`'d after this instruction.
+- `call foo`: starts executing instructions defined by the function `foo`
```

Comment:
Missing a period (the other bullet in this list has one).

Reply:
```suggestion
- `call foo`: starts executing instructions defined by the function `foo`.
```